### ISSUE TITLE
CUDA overrides somehow screwing up customized printing? 

### ISSUE THREAD

**Issue by marius311** [JuliaGPU/CUDA.jl#1055]
This is really bizarre but reproducible. It seems loading CUDA screws up overloaded printing of types. Eg:

```julia
julia> using CUDA

julia> struct Foo{T} <: AbstractVector{T} end

julia> Base.size(::Foo) = (0,)

julia> Base.show_datatype(io::IO, t::Type{<:Foo}) = print(io, "custom")

julia> Foo{Float32}()
0-element Foo{Float32}
```

whereas if in a fresh session you omit `using CUDA`, you (correctly) get `0-element custom`.

I've narrowed it down to that commenting out the `eval(overrides)` in `CUDA.__init__()` fixes it, but can't get any further. 

This is Julia 1.6.1, CUDA 3.3.4 or master.

**Comment by maleadt** [comment]
Can't reproduce; didn't you mention on Slack you could only reproduce it yourself on a specific Jupyter instance?

**Comment by marius311** [comment]
That's what I thought at first, but then realized I actually could reproduce in the REPL, at least on two systems I tried. Will try to think what else could be affecting reproducibility...

**Comment by maleadt** [comment]
You can also try commenting-out the `eval(overrides)` and then doing an `eval(CUDA.overrides.args[i])` for each override to see which one messes with the displaying.

**Comment by maleadt** [comment]
Also try with Julia 1.7 -- use of method override tables there should not mess with existing methods as much.

**Comment by marius311** [comment]
On 1.7 I'm not seeing this. 

I also reduced it further and discovered that 1) the key piece is whether overrides are defined before or after the custom `show_datatype`, and 2) its only when actually printing to the REPL, not when calling `Base.show` manually (which I think is what is used to print?; this makes it hard to systematically check for which of the many overrides is the offending one..) 

Here it is triggered:

```julia
julia> using CUDA # with overrides commented out

julia> struct Foo{T} <: AbstractVector{T} end

julia> CUDA.eval(CUDA.overrides)

julia> Base.show_datatype(io::IO, t::Type{<:Foo}) = print(io, "custom")

julia> Foo{Float32}
Foo{Float32}

julia> Base.show(stdout, Foo{Float32})
custom
```

and not triggered

```julia
julia> using CUDA # with overrides commented out

julia> struct Foo{T} <: AbstractVector{T} end

julia> Base.show_datatype(io::IO, t::Type{<:Foo}) = print(io, "custom")

julia> CUDA.eval(CUDA.overrides)

julia> Foo{Float32}
custom

julia> Base.show(stdout, Foo{Float32})
custom
```

This must be some world-age age type thing? Like something in the overrides is forcing a compilation of the generic `show_datatype`, at least when used for REPL printing, which then never gets updated for some reason once I define the custom one? I looked through the overrides but nothing stuck out as interacting with `show_datatype` in any way. It is weird that you're unable to reproduce (I've checked I have nothing in my in startup.jl, its Linux, can't think what else it could be...).

**Comment by maleadt** [comment]
Admittedly, we are messing with world ages a little on 1.6, so I'm glad it only seems to affect a `show` method here. Since we have a better solution on 1.7, which doesn't introduce this issue, I don't think this is worth the time to solve this?

**Comment by marius311** [comment]
Yea, I just realized that a workaround for this is to always do `using MyPkg, CUDA` as opposed to `using CUDA, MyPkg`, and since its fixed on 1.7 anyway, I'd say this is too small fish to try, closing.
